### PR TITLE
feat: handle message send failure

### PR DESCRIPTION
**Pull request by trOnk12**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- When sending a message fails in any step we want to change the status to the FAILED one

### Solutions

- change the status to the FAILED on onFailure of the function wrapping all the steps

### Testing

-Unit test

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+                .wasInvoked(Times(1))
+
+            assertIs<Either.Left<Unit>>(result)
+        }
+    }
+
+    @Test
+    fun givenGettingConversationRecipientsFails_WhenSendingOutgoingMessage_ThenReturnFailureAndSetMessageStatusToFailed() {
+        runTest {
+            //given
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:119]:
```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+                .suspendFunction(messageRepository::updateMessageStatus)
+                .with(eq(MessageEntity.Status.FAILED), anything(), anything())
+                .wasInvoked(Times(1))
+
+            assertIs<Either.Left<Unit>>(result)
+        }
+    }
+
+    @Test
+    fun givenGettingConversationRecipientsFails_WhenSendingOutgoingMessage_ThenReturnFailureAndSetMessageStatusToFailed() {
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:117]: It seems like this test is the same as the one above. Just with a different name. Should the given below be different?

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        runTest {
+            //given
+            //given
+            setupGivenSuccessResults(
+                getMessageById = true,
+                getConversationProtocol = false,
+                getConversationsRecipient = true,
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:124]:
```suggestion
                getConversationProtocol = true,
                getConversationsRecipient = false,
```

I guess it was supposed to be like this?

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            val result = messageSender.trySendingOutgoingMessageById(ConversationId("testId", "testDomain"), "testId")
+            //then
+            verify(messageRepository)
+                .suspendFunction(messageRepository::updateMessageStatus)
+                .with(eq(MessageEntity.Status.FAILED), anything(), anything())
+                .wasInvoked(Times(1))
+
+            assertIs<Either.Left<Unit>>(result)
+        }
+    }
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:167]: This test seems to be a copy of the one before, just with a different name.

I suggest:

```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+    @Test
+    fun givenPreparingRecipentsForNewOutgoingMessageFails_WhenSendingOutgoingMessage_ThenReturnFailureAndSetMessageStatusToFailed() {
+        runTest {
+            //given
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:172]:
```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+                .wasInvoked(Times(1))
+
+            assertIs<Either.Left<Unit>>(result)
+        }
+    }
+
+    @Test
+    fun givenCreatingOutgoingEnvelopeFails_WhenSendingOutgoingMessage_ThenReturnFailureAndSetMessageStatusToFailed() {
+        runTest {
+            //given
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:198]:
```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+                .wasInvoked(Times(1))
+
+            assertIs<Either.Left<Unit>>(result)
+        }
+    }
+
+    @Test
+    fun givenSendingEnvelopeFails_WhenSendingOutgoingMessage_ThenReturnFailureAndSetMessageStatusToFailed() {
+        runTest {
+            //given
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:224]:
```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+                .wasInvoked(Times(1))
+
+            assertIs<Either.Left<Unit>>(result)
+        }
+    }
+
+    @Test
+    fun givenUpdatingMessageStatusToSuccessFails_WhenSendingOutgoingMessage_ThenReturnFailureAndSetMessageStatusToFailed() {
+        runTest {
+            //given
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:250]:
```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+            assertIs<Either.Left<Unit>>(result)
+            //when
+        }
+    }
+
+    @Test
+    fun testing3() {
+        runTest {
+            //given
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:146]:
```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+            //when
+            val result = messageSender.trySendingOutgoingMessageById(ConversationId("testId", "testDomain"), "testId")
+            //then
+            verify(messageRepository)
+                .suspendFunction(messageRepository::updateMessageStatus)
+                .with(eq(MessageEntity.Status.FAILED), anything(), anything())
+                .wasInvoked(Times(1))
+
+            assertIs<Either.Left<Unit>>(result)
+            //when
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:139]:
```suggestion
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SendTextMessageUseCaseTest.kt [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+//
+//        verify(messageSender)
+//            .suspendFunction(messageSender::trySendingOutgoingMessageById)
+//            .with(anything(), anything())
+//            .wasInvoked(Times(1))
+//
+//        assertIs<SendTextMessageResult.Success>(result)
+//    }
+//
+//}
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SendTextMessageUseCaseTest.kt:229]: Is this intentional? Why commenting the tests out?

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Simple and elegant handling of failure :) 

Just some tiny C&P mistakes when elaborating the tests and a question about commenting out the `SendTextMessageUseCaseTest`.

Also, in order to reduce identation in tests:

```kotlin
@Test
fun myTest() {
    runTest {
    
    }
}
```

Can be replaced with:

```kotlin
@Test
fun myTest() = runTest {

}
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt (hunk 2 above)
Comment (trOnk12) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:117]: true, did not notice it, thanks

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SendTextMessageUseCaseTest.kt (hunk 11 above)
Comment (trOnk12) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SendTextMessageUseCaseTest.kt:229]: this is fro mthe previous take on this, those tests are not valid anymore , will remove it, thanks

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+        }
+    }
+
+    @Test
+    fun givenGettingConversationRecipientsFails_WhenSendingOutgoingMessage_ThenReturnFailureAndSetMessageStatusToFailed() {
+        runTest {
+            //given
+            setupGivenSuccessResults(
+                getMessageById = true,
+                getConversationProtocol = false,
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageSenderTest.kt:122]:
```suggestion
               getConversationProtocol = true,
```

And this one be true :P

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Oooone more tiny detail :P

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt [hunk 13]
```diff
@@ -28,8 +28,12 @@ interface MessageRepository {
...
     suspend fun softDeleteMessage(messageUuid: String, conversationId: ConversationId): Either<CoreFailure, Unit>
     suspend fun hideMessage(messageUuid: String, conversationId: ConversationId): Either<CoreFailure, Unit>
-    suspend fun markMessageAsSent(conversationId: ConversationId, messageUuid: String): Either<CoreFailure, Unit>
-    suspend fun updateMessageDate(conversationId: ConversationId, messageUuid: String, date: String): Either<CoreFailure, Unit>
+    suspend fun updateMessageStatus(
+        messageStatus: MessageEntity.Status,
+        conversationId: ConversationId,
+        messageUuid: String
+    ): Either<CoreFailure, Unit>
+suspend fun updateMessageDate(conversationId: ConversationId, messageUuid: String, date: String): Either<CoreFailure, Unit>
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt:36]:
```suggestion
    suspend fun updateMessageDate(conversationId: ConversationId, messageUuid: String, date: String): Either<CoreFailure, Unit>
```

Unintentional I guess?

Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/MessageRepository.kt]: yes no idea what is happening xd

**Review by vitorhugods — APPROVED** [review]
🚀🚀🚀🚀🚀🚀

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt [hunk 14]
```diff
@@ -95,14 +99,15 @@ class MessageSenderImpl(
...
-            // TODO handle mls-stale-message
-            messageRepository.sendMLSMessage(conversationId, mlsMessage).map {
+    ): Either<CoreFailure, String> =
+        suspending {
+            mlsMessageCreator.createOutgoingMLSMessage(groupId, message).flatMap { mlsMessage ->
+                // TODO handle mls-stale-message
+                messageRepository.sendMLSMessage(conversationId, mlsMessage).map {
                 message.date //TODO return actual server time from the response
             }
+            }
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt:109]: Can you fix this weird indentation?

**Review by gongracr — APPROVED** [review]
Looking really solid! ⚒️ 🪨

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt [hunk 15]
```diff
@@ -57,10 +59,15 @@ class MessageSenderImpl(
...
-                    val millisDiff = message.date.toInstant().until(messageRemoteTime.toInstant(), DateTimeUnit.MILLISECOND)
-                    // this should make sure that pending messages are ordered correctly after one of them is sent
-                    messageRepository.updatePendingMessagesAddMillisToDate(conversationId, millisDiff)
+                        .flatMap {
+                            messageRepository.updateMessageStatus(MessageEntity.Status.SENT, conversationId, message.id)
+                        }.flatMap {
+                            // this should make sure that pending messages are ordered correctly after one of them is sent
+                            messageRepository.updatePendingMessagesAddMillisToDate(
+                                conversationId,
+                                timeParser.calculateMillisDifference(message.date, messageRemoteTime)
```
Comment (trOnk12) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt:68]: this line 
`val millisDiff = message.date.toInstant().until(messageRemoteTime.toInstant(), DateTimeUnit.MILLISECOND)`

was requiring to pass a valid date to messageRepository.getMessageById when mocking the messageRepository for unit tests, also it required to some other function to return a valid messageRemoteTime format. 

I solved it by creating an interface that we can mock, this method :
val millisDiff = message.date.toInstant().until(messageRemoteTime.toInstant(), DateTimeUnit.MILLISECOND)

could throw an error, when it does, the app crashes, but we both discussed with Michal that it should never happen we think

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/util/TimeParser.kt [hunk 16]
```diff
@@ -4,6 +4,7 @@ import kotlinx.datetime.DateTimeUnit
 import kotlinx.datetime.toInstant
 import kotlinx.datetime.until
 
+//TODO: we need to think if this TimeParser should return an either, should we catch the exception ?
 interface TimeParser {
     fun calculateMillisDifference(time1: String, time2: String): Long
```
Comment (saleniuk) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/util/TimeParser.kt:9]: I would add the comment to make sure everyone using it knows that the datetime we use needs to be the ISO-8601 format (e.g. `YYYY-MM-DDTHH:mm:ss.SSSZ`)

Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/util/TimeParser.kt:9]: will do :)

**Review by saleniuk — APPROVED** [review]

**Review by vitorhugods — APPROVED** [review]
Nice job.

Just makes me think that MessageSender is growing a bit too much :P 

Perhaps we can split it in the future. Separate AVS from Proteus, for example.

**Pull request merged**